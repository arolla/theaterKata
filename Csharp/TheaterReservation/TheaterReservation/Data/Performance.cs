using String = System.String;

namespace TheaterReservation.Data
{
    public class Performance
    {
        public Int64 id;
     
        public String play; // "The CICD - Corneille", "Les fourberies de Scala - Molière"

        public DateTime startTime;

        public DateTime endTime;

        public String performanceNature; // can be "PREVIEW", "PREMIERE", etc
    }
}
