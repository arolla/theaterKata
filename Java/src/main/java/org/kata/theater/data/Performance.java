package org.kata.theater.data;

import java.time.LocalDateTime;

public class Performance {

    public long id;
    public String play; // "The CICD - Corneille", "Les fourberies de Scala - Moli√®re"

    public LocalDateTime startTime;

    public LocalDateTime endTime;

    public String performanceNature; // can be "PREVIEW", "PREMIERE", etc
}
