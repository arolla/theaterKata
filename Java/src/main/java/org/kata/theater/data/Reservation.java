package org.kata.theater.data;

import java.util.Arrays;

public class Reservation {

    private Long reservationId;

    private Long performanceId;

    private String status;

    private String[] seats;

    public Long getReservationId() {
        return reservationId;
    }

    public void setReservationId(Long reservationId) {
        this.reservationId = reservationId;
    }

    public Long getPerformanceId() {
        return performanceId;
    }

    public void setPerformanceId(Long performanceId) {
        this.performanceId = performanceId;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String[] getSeats() {
        return seats;
    }

    public void setSeats(String[] seats) {
        this.seats = seats;
    }

    @Override
    public String toString() {
        return "Reservation{" +
                "reservationId=" + reservationId +
                ", status='" + status + '\'' +
                ", seats=" + Arrays.toString(seats) +
                '}';
    }
}
