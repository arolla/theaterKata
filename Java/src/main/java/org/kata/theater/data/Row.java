package org.kata.theater.data;

public class Row {

    private Seat[] seats;

    public Row(Seat[] seats) {
        this.seats = seats;
    }

    public Seat[] getSeats() {
        return seats;
    }

    public void setSeats(Seat[] seats) {
        this.seats = seats;
    }
}
