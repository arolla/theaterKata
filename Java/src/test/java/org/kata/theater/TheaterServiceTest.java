package org.kata.theater;

import org.approvaltests.Approvals;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.kata.theater.data.Performance;

import java.time.LocalDate;
import java.time.Month;
import java.util.List;

class TheaterServiceTest {

    private TheaterService theaterService;



    @BeforeEach
    void setUp() {
        theaterService = new TheaterService();
    }

    @Test
    void reserve_once_on_premiere_performance() {
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation = theaterService.reservation(1L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);

        // TODO : add testing for reserved seat references
        Assertions.assertThat(ReservationService.findReservation(123460)).isNotNull();

    }

    @Test
    void reserve_once_on_premiere_performance_with_premium_category() {
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation = theaterService.reservation(1L, 4, "PREMIUM",
                performance);
        Approvals.verify(reservation);

        Assertions.assertThat(ReservationService.findReservation(123459)).isNotNull();
    }

    @Test
    void cancel_then_reserve_on_premiere_performance_with_standard_category() {
        theaterService.cancelReservation("123456", 1L, List.of("B2"));
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation = theaterService.reservation(1L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);

        Assertions.assertThat(ReservationService.findReservation(123457)).isNotNull();
    }

    @Test
    void reserve_twice_on_premiere_performance() {
        Performance performance = new Performance();
        performance.id = 1L;
        performance.play = "The CICD by Corneille";
        performance.startTime = LocalDate.of(2023, Month.APRIL, 22).atTime(21, 0);
        performance.performanceNature = "PREMIERE";
        String reservation1 = theaterService.reservation(1L, 4, "STANDARD",
                performance);
        String reservation2 = theaterService.reservation(1L, 5, "STANDARD",
                performance);
        Approvals.verify(reservation2);

        Assertions.assertThat(ReservationService.findReservation(123461)).isNotNull();
    }

    @Test
    void reservation_failed_on_preview_performance() {
        Performance performance = new Performance();
        performance.id = 2L;
        performance.play = "Les fourberies de Scala - Molière";
        performance.startTime = LocalDate.of(2023, Month.MARCH, 21).atTime(21, 0);
        performance.performanceNature = "PREVIEW";

        String reservation = theaterService.reservation(2L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);

        Assertions.assertThat(ReservationService.findReservation(123458)).isNotNull();
    }

    @Test
    void reservation_failed_on_premiere_performance() {
        Performance performance = new Performance();
        performance.id = 3L;
        performance.play = "DOM JSON - Molière";
        performance.startTime = LocalDate.of(2023, Month.MARCH, 21).atTime(21, 0);
        performance.performanceNature = "PREMIERE";

        String reservation = theaterService.reservation(2L, 4, "STANDARD",
                performance);
        Approvals.verify(reservation);

        Assertions.assertThat(ReservationService.findReservation(123456)).isNotNull();
    }

}
