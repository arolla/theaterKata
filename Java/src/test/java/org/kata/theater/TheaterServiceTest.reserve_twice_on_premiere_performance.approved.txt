<reservation>
	<performance>
		<play>The CICD by Corneille</play>
		<date>2023-04-22</date>
		<time>21:00</time>
	</performance>
	<reservationId>123462</reservationId>
	<reservationStatus>FULFILLABLE</reservationStatus>
		<seats>
			<seat>
				<id>C1</id>
				<category>STANDARD</category>
			</seat>
			<seat>
				<id>C2</id>
				<category>STANDARD</category>
			</seat>
			<seat>
				<id>C3</id>
				<category>STANDARD</category>
			</seat>
			<seat>
				<id>C4</id>
				<category>STANDARD</category>
			</seat>
			<seat>
				<id>C5</id>
				<category>STANDARD</category>
			</seat>
		</seats>
	<seatCategory>STANDARD</seatCategory>
	<totalAmountDue>115.50â‚¬</totalAmountDue>
</reservation>
